File: common/src/main/java/com/leo/commonutil/app/SystemUtils.java
package com.leo.commonutil.app;

import android.Manifest;
import android.app.Activity;
import android.app.ActivityManager;
import android.app.KeyguardManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.content.res.TypedArray;
import android.graphics.Rect;
import android.net.Uri;
import android.os.Build;
import android.os.Looper;
import android.os.PowerManager;
import android.os.Process;
import android.provider.Settings;
import android.support.annotation.ColorInt;
import android.support.annotation.ColorRes;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresPermission;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.FileProvider;
import android.telephony.TelephonyManager;
import android.text.TextUtils;
import android.util.DisplayMetrics;
import android.view.KeyEvent;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;
import android.view.inputmethod.InputMethodManager;
import android.widget.EditText;

import com.leo.commonutil.callback.OnEditTextClearFocusCallback;
import com.leo.commonutil.storage.SPHelp;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.io.UnsupportedEncodingException;
import java.lang.ref.WeakReference;
import java.lang.reflect.Method;
import java.util.List;
import java.util.UUID;

public final class SystemUtils {
    private SystemUtils() {
    }

    public static int getKeyboardHeight(Activity paramActivity) {
        int height = SystemUtils.getScreenHeight(paramActivity)
                - SystemUtils.getStatusBarHeight(paramActivity)
                - SystemUtils.getAppHeight(paramActivity);
        if (height == 0) {
            height = SPHelp.getInstance().getInt(paramActivity, "KeyboardHeight",
                    dp2px(paramActivity, 295));//295dp-787为默认软键盘高度 基本差不离
        } else {
            SPHelp.getInstance().put(paramActivity, "KeyboardHeight", height);
        }
        return height;
    }

    /**
     * 得到设备屏幕的高度
     **/
    public static int getScreenHeight(Context context) {
        DisplayMetrics dm = context.getResources().getDisplayMetrics();
        return dm.heightPixels;
    }

    /**
     * 得到设备屏幕的宽度
     */
    public static int getScreenWidth(Context context) {
        DisplayMetrics dm = context.getResources().getDisplayMetrics();
        return dm.widthPixels;
    }

    /**
     * statusBar高度
     **/
    public static int getStatusBarHeight(Activity paramActivity) {
        Rect localRect = new Rect();
        paramActivity.getWindow().getDecorView().getWindowVisibleDisplayFrame(localRect);
        return localRect.top;

    }

    /**
     * 可见屏幕高度
     **/
    public static int getAppHeight(Activity paramActivity) {
        Rect localRect = new Rect();
        paramActivity.getWindow().getDecorView().getWindowVisibleDisplayFrame(localRect);
        return localRect.height();
    }

    // below actionbar, above softkeyboard
    public static int getAppContentHeight(Activity paramActivity) {
        return SystemUtils.getScreenHeight(paramActivity) - SystemUtils.getStatusBarHeight(paramActivity)
                - SystemUtils.getActionBarHeight(paramActivity) - SystemUtils.getKeyboardHeight(paramActivity);
    }

    /**
     * 获取actiobar高度
     **/
    public static int getActionBarHeight(Activity paramActivity) {
        if (true) {
            return SystemUtils.dp2px(paramActivity, 48);
        }
        int[] attrs = new int[]{android.R.attr.actionBarSize};
        TypedArray ta = paramActivity.obtainStyledAttributes(attrs);
        return ta.getDimensionPixelSize(0, SystemUtils.dp2px(paramActivity, 48));
    }

    /**
     * 根据手机的分辨率从 dp 的单位 转成为 px(像素)
     */
    public static int dp2px(Context context, float dpValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (dpValue * scale + 0.5f);
    }

    /**
     * 根据手机的分辨率从 px(像素) 的单位 转成为 dp
     */
    public static int px2dp(Context context, float pxValue) {
        final float scale = context.getResources().getDisplayMetrics().density;
        return (int) (pxValue / scale + 0.5f);
    }

    /**
     * 键盘是否在显示
     **/
    public static boolean isKeyBoardShow(Activity paramActivity) {
        int height = SystemUtils.getScreenHeight(paramActivity)
                - SystemUtils.getStatusBarHeight(paramActivity)
                - SystemUtils.getAppHeight(paramActivity);
        return height != 0;
    }

    /**
     * 显示键盘
     **/
    public static void showKeyBoard(Context mContext, final EditText paramEditText) {
        WeakReference<Context> weakReference = new WeakReference<>(mContext);
        paramEditText.requestFocus();
        paramEditText.post(() -> {
            Context context = weakReference.get();
            if (null == context) {
                return;
            }
            InputMethodManager imm = (InputMethodManager) context.getApplicationContext()
                    .getSystemService(Context.INPUT_METHOD_SERVICE);
            if (null != imm) {
                imm.showSoftInput(paramEditText, 0);
            }
        });
    }

    /**
     * 关闭软键盘
     *
     * @param mEditText 获得焦点的输入框
     * @param mContext  上下文
     */
    public static void hideSoftKeyboard(EditText mEditText, Context mContext) {
        try {
            mEditText.clearFocus();
            InputMethodManager imm = (InputMethodManager) mContext
                    .getSystemService(Context.INPUT_METHOD_SERVICE);
            if (null != imm) {
                imm.hideSoftInputFromWindow(mEditText.getWindowToken(), 0);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public static void hideSoftKeyboard(Activity activity) {
        try {
            InputMethodManager inputMethodManager = (InputMethodManager) activity
                    .getSystemService(Activity.INPUT_METHOD_SERVICE);
            if (null != inputMethodManager) {
                View currentFocus = activity.getCurrentFocus();
                if (null != currentFocus) {
                    inputMethodManager.hideSoftInputFromWindow(currentFocus.getWindowToken(), 0);
                }
            }
        } catch (NullPointerException e) {
            e.printStackTrace();
        }
    }

    /**
     * 设置外部点击隐藏软键盘,传入根布局.
     *
     * @param context
     * @param view
     */
    public static void setCanceledOnTouchOutsideET(final Context context, View view) {
        setCanceledOnTouchOutsideET(context, view, null);
    }

    public static void setCanceledOnTouchOutsideET(final Context context, View view,
                                                   @Nullable final OnEditTextClearFocusCallback onEditTextClearFocusCallback) {
        //Set up touch listener for non-text box views to hide keyboard.
        if (!(view instanceof EditText)) {
            view.setOnTouchListener((v, event) -> {
                SystemUtils.hideSoftKeyboard((Activity) context);
                if (null != onEditTextClearFocusCallback) {
                    onEditTextClearFocusCallback.editTextClearFocusListener();
                }
                return false;
            });
        }

        //If a layout container, iterate over children and seed recursion.
        if (view instanceof ViewGroup) {
            for (int i = 0; i < ((ViewGroup) view).getChildCount(); i++) {
                View innerView = ((ViewGroup) view).getChildAt(i);
                setCanceledOnTouchOutsideET(context, innerView, onEditTextClearFocusCallback);
            }
        }
    }

    /**
     * 设置scrollview中的Edittext可滚动
     *
     * @param mEditEt
     * @param event
     */
    public static void setScrollEditText(EditText mEditEt, MotionEvent event) {
        if (canVerticalScroll(mEditEt)) {
            mEditEt.getParent().requestDisallowInterceptTouchEvent(true);//告诉父view，我的事件自己处理
            if (event.getAction() == MotionEvent.ACTION_UP) {
                mEditEt.getParent().requestDisallowInterceptTouchEvent(false);//告诉父view，你可以处理了
            }
        }
    }

    /**
     * EditText竖直方向是否可以滚动
     *
     * @param editText 需要判断的EditText
     * @return true：可以滚动   false：不可以滚动
     */
    public static boolean canVerticalScroll(EditText editText) {
        //滚动的距离
        int scrollY = editText.getScrollY();
        //控件内容的总高度
        int scrollRange = editText.getLayout().getHeight();
        //控件实际显示的高度
        int scrollExtent = editText.getHeight()
                - editText.getCompoundPaddingTop()
                - editText.getCompoundPaddingBottom();
        //控件内容总高度与实际显示高度的差值
        int scrollDifference = scrollRange - scrollExtent;
        if (scrollDifference == 0) {
            return false;
        }
        return (scrollY > 0) || (scrollY < scrollDifference - 1);
    }

    /**
     * 逐个删除EditText的输入元素
     *
     * @param editText
     */
    public static void editTextDel(EditText editText) {
        int keyCode = KeyEvent.KEYCODE_DEL;
        KeyEvent keyEventDown = new KeyEvent(KeyEvent.ACTION_DOWN, keyCode);
        KeyEvent keyEventUp = new KeyEvent(KeyEvent.ACTION_UP, keyCode);
        editText.onKeyDown(keyCode, keyEventDown);
        editText.onKeyUp(keyCode, keyEventUp);
    }

    /**
     * 是否启用了代理
     *
     * @param context
     * @return
     */
    public static boolean isWifiProxy(Context context) {
        final boolean IS_ICS_OR_LATER = Build.VERSION.SDK_INT >= Build.VERSION_CODES.ICE_CREAM_SANDWICH;
        String proxyAddress;
        int proxyPort;
        if (IS_ICS_OR_LATER) {
            proxyAddress = System.getProperty("http.proxyHost");
            String portStr = System.getProperty("http.proxyPort");
            proxyPort = Integer.parseInt((portStr != null ? portStr : "-1"));
        } else {
            proxyAddress = android.net.Proxy.getHost(context);
            proxyPort = android.net.Proxy.getPort(context);
        }
        return (!TextUtils.isEmpty(proxyAddress)) && (proxyPort != -1);
    }

    /**
     * 移除所有代理
     */
    public static void removeAllWifiProxy() {
        System.getProperties().remove("http.proxyHost");
        System.getProperties().remove("http.proxyPort");
        System.getProperties().remove("https.proxyHost");
        System.getProperties().remove("https.proxyPort");
    }

    /**
     * 更改状态栏
     *
     * @param activity
     * @param color
     */
    public static void setStatuBarColorRes(Activity activity, @ColorRes int color) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && null != activity) {
            Window window = activity.getWindow();
            if (null != window) {
                window.setStatusBarColor(activity.getResources().getColor(color));
            }
        }
    }

    public static void setStatuBarColor(Activity activity, @ColorInt int color) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP && null != activity) {
            Window window = activity.getWindow();
            if (null != window) {
                window.setStatusBarColor(color);
            }
        }
    }

    /**
     * 安装APK
     */
    public static void installApk(Context context, File file) {
        Intent intent = new Intent(Intent.ACTION_VIEW);
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            intent.setDataAndType(Uri.fromFile(file), "application/vnd.android.package-archive");
        } else {
            Uri apkUri = FileProvider.getUriForFile(context, context.getPackageName() + ".fileprovider", file);
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            intent.setDataAndType(apkUri, "application/vnd.android.package-archive");
        }
        context.startActivity(intent);
    }

    /**
     * 通知Media扫描
     *
     * @param context
     * @param file
     */
    public static void notifyMediaScan(Context context, File file) {
        Intent intent = new Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE);//发送更新图片信息广播
        Uri uri = Uri.fromFile(file);
        intent.setData(uri);
        context.getApplicationContext().sendBroadcast(intent);
    }

    /**
     * 权限检查
     *
     * @param context     上下文
     * @param permissions 权限数组
     * @return 返回检查结果
     */
    public static boolean checkPermissions(Context context, String... permissions) {
        if (null == permissions || permissions.length == 0)
            return true;
        boolean isGranted = true;
        for (String permission : permissions) {
            isGranted = ActivityCompat.checkSelfPermission(context, permission) == PackageManager.PERMISSION_GRANTED;
            if (!isGranted)
                break;
        }
        return isGranted;
    }

    private static long lastClickTime;

    public static boolean isFastClick() {//这里设置的时间间隔是2s-判断是否频繁点击
        return isFastClick(1000);
    }

    /**
     * @param intervalTime 时间间隔
     * @return
     */
    public static boolean isFastClick(int intervalTime) {//判断是否频繁点击
        long time = System.currentTimeMillis();
        long timeD = time - lastClickTime;
        if (0 < timeD && timeD < intervalTime) {
            return true;
        }
        lastClickTime = time;
        return false;
    }

    /**
     * 是否主进程
     *
     * @param context
     * @return
     */
    public static boolean isMainProcess(Context context) {
        ActivityManager am = ((ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE));
        if (null != am) {
            List<ActivityManager.RunningAppProcessInfo> processInfos = am.getRunningAppProcesses();
            String mainProcessName = context.getPackageName();
            int myPid = Process.myPid();
            for (ActivityManager.RunningAppProcessInfo info : processInfos) {
                if (info.pid == myPid && mainProcessName.equals(info.processName)) {
                    return true;
                }
            }
        }
        return false;
    }

    /**
     * 获取系统基带
     *
     * @return
     */
    public static String getSystemBaseband() {
        try {
            Class cl = Class.forName("android.os.SystemProperties");
            Object invoker = cl.newInstance();
            Method m = cl.getMethod("get", new Class[]{String.class, String.class});
            Object result = m.invoke(invoker, new Object[]{"gsm.version.baseband", ""});
            return (String) result;
        } catch (Exception e) {

        }
        return "";
    }

    /**
     * 判断是否打开了允许虚拟位置
     */
    public static boolean isAllowMockLocation(final Activity context) {
        boolean isOpen = Settings.Secure.getInt(context.getContentResolver(),
                Settings.Secure.ALLOW_MOCK_LOCATION, 0) != 0;
        /*
         * 该判断API是androidM以下的API,由于Android M中已经没有了关闭允许模拟位置的入口,所以这里一旦检测到开启了模拟位置,并且是android M以上,则
         * 默认设置为未有开启模拟位置
         * */
        if (isOpen && Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP_MR1) {
            isOpen = false;
        }
        return isOpen;
    }

    /**
     * 获取进程号对应的进程名
     *
     * @param pid 进程号
     * @return 进程名
     */
    public static String getProcessName(int pid) {
        BufferedReader reader = null;
        try {
            reader = new BufferedReader(new FileReader("/proc/" + pid + "/cmdline"));
            String processName = reader.readLine();
            if (!TextUtils.isEmpty(processName)) {
                processName = processName.trim();
            }
            return processName;
        } catch (Throwable throwable) {
            throwable.printStackTrace();
        } finally {
            try {
                if (reader != null) {
                    reader.close();
                }
            } catch (IOException exception) {
                exception.printStackTrace();
            }
        }
        return null;
    }

    /**
     * 是否主线程
     *
     * @return
     */
    public static boolean isMainThread() {
        return Looper.getMainLooper().getThread() == Thread.currentThread();
    }

    /**
     * 获取服务是否开启
     */
    public static boolean isServiceRunning(Context context, String className) {
        ActivityManager activityManager = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        //获取正在运行的服务
        List<ActivityManager.RunningServiceInfo> runningServices = activityManager.getRunningServices(1000);//maxNum 返回正在运行的服务的上限个数,最多返回多少个服务
        for (ActivityManager.RunningServiceInfo runningServiceInfo : runningServices) {
            ComponentName service = runningServiceInfo.service;
            //获取正在运行的服务的全类名
            String className2 = service.getClassName();
            //将获取到的正在运行的服务的全类名和传递过来的服务的全类名比较,一直表示服务正在运行  返回true,不一致表示服务没有运行  返回false
            if (className.equals(className2)) {
                return true;
            }
        }
        return false;
    }

    /**
     * 判断当前App是否处于后台
     *
     * @param context
     * @return
     */
    public static boolean isApplicationBackground(final Context context) {
        ActivityManager am = (ActivityManager) context.getSystemService(Context.ACTIVITY_SERVICE);
        List<ActivityManager.RunningTaskInfo> tasks = am.getRunningTasks(1);
        if (!tasks.isEmpty()) {
            ComponentName topActivity = tasks.get(0).topActivity;
            if (!topActivity.getPackageName().equals(context.getPackageName())) {
                return true;
            }
        }
        return false;
    }

    /**
     * 唤醒屏幕并解锁
     *
     * @param context
     */
    @RequiresPermission(Manifest.permission.DISABLE_KEYGUARD)
    public static void wakeUpAndUnlock(Context context) {
        try {
            KeyguardManager km = (KeyguardManager) context.getSystemService(Context.KEYGUARD_SERVICE);
            KeyguardManager.KeyguardLock kl = km.newKeyguardLock("unLock");
            //解锁
            kl.disableKeyguard();
            //获取电源管理器对象
            PowerManager pm = (PowerManager) context.getSystemService(Context.POWER_SERVICE);
            //获取PowerManager.WakeLock对象,后面的参数|表示同时传入两个值,最后的是LogCat里用的Tag
            PowerManager.WakeLock wl = pm.newWakeLock(PowerManager.ACQUIRE_CAUSES_WAKEUP | PowerManager.SCREEN_DIM_WAKE_LOCK, "bright");
            //点亮屏幕
            wl.acquire();
            //释放
            wl.release();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    /**
     * 获取UDID
     */
    protected static final String PREFS_FILE = "gank_device_id.xml";
    protected static final String PREFS_DEVICE_ID = "gank_device_id";
    protected static String uuid;

    @RequiresPermission(android.Manifest.permission.READ_PHONE_STATE)
    public static String getUDID(Context context) {
        if (uuid == null) {
            synchronized (PREFS_FILE) {
                if (uuid == null) {
                    final SharedPreferences prefs = context.getApplicationContext().getSharedPreferences(PREFS_FILE, Context.MODE_PRIVATE);
                    final String id = prefs.getString(PREFS_DEVICE_ID, null);
                    if (id != null) {
                        uuid = id;
                    } else {
                        final String androidId = Settings.Secure.getString(context.getApplicationContext().getContentResolver(), Settings.Secure.ANDROID_ID);
                        try {
                            if (!"9774d56d682e549c".equals(androidId)) {
                                uuid = UUID.nameUUIDFromBytes(androidId.getBytes("utf8")).toString();
                            } else {
                                if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) == PackageManager.PERMISSION_GRANTED) {
                                    final String deviceId = ((TelephonyManager) context.getApplicationContext().getSystemService(Context.TELEPHONY_SERVICE)).getDeviceId();
                                    uuid = deviceId != null ? UUID.nameUUIDFromBytes(deviceId.getBytes("utf8")).toString() : UUID.randomUUID().toString();
                                } else {
                                    uuid = UUID.randomUUID().toString();
                                }
                            }
                        } catch (UnsupportedEncodingException e) {
                            throw new RuntimeException(e);
                        }
                        prefs.edit().putString(PREFS_DEVICE_ID, uuid).commit();
                    }
                }
            }
        }
        return uuid;
    }
}

File: system/src/main/java/com/leo/system/util/SystemUtils.kt
package com.leo.system.util

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.app.ActivityManager
import android.app.KeyguardManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageInfo
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Looper
import android.os.PowerManager
import android.os.Process
import android.provider.Settings
import android.telephony.TelephonyManager
import android.text.TextUtils
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.EditText
import androidx.annotation.RequiresPermission
import androidx.core.app.ActivityCompat
import androidx.core.content.FileProvider
import com.leo.system.database.callback.IETClearFocus
import com.leo.system.context.ContextHelper
import com.leo.system.context.ContextHelper.context
import java.io.BufferedReader
import java.io.File
import java.io.FileReader
import java.io.UnsupportedEncodingException
import java.util.*

object SystemUtils {

    /**
     * 设置外部点击隐藏软键盘,传入根布局.
     *
     * @param context
     * @param view
     */
    fun setCancelOnTouchOutsideET(context: Context, view: View) {
        setCancelOnTouchOutsideET(context, view, null)
    }

    @SuppressLint("ClickableViewAccessibility")
    fun setCancelOnTouchOutsideET(context: Context, view: View,
                                  iETClearFocus: IETClearFocus?) {
        //Set up touch listener for non-text box views to hide keyboard.
        if (view !is EditText) {
            view.setOnTouchListener { v: View?, event: MotionEvent? ->
                KeyboardHelper.hideSoftKeyboard(context as Activity)
                iETClearFocus?.onClearFocus()
                false
            }
        }

        //If a layout container, iterate over children and seed recursion.
        if (view is ViewGroup) {
            for (i in 0 until view.childCount) {
                val innerView = view.getChildAt(i)
                setCancelOnTouchOutsideET(context, innerView, iETClearFocus)
            }
        }
    }

    /**
     * 设置scrollview中的Edittext可滚动
     *
     * @param mEditEt
     * @param event
     */
    fun setScrollEditText(mEditEt: EditText, event: MotionEvent) {
        if (canVerticalScroll(mEditEt)) {
            // 告诉父view，我的事件自己处理
            mEditEt.parent.requestDisallowInterceptTouchEvent(true)
            if (event.action == MotionEvent.ACTION_UP) {
                // 告诉父view，你可以处理了
                mEditEt.parent.requestDisallowInterceptTouchEvent(false)
            }
        }
    }

    /**
     * EditText竖直方向是否可以滚动
     *
     * @param editText 需要判断的EditText
     * @return true：可以滚动   false：不可以滚动
     */
    fun canVerticalScroll(editText: EditText): Boolean {
        // 滚动的距离
        val scrollY = editText.scrollY
        // 控件内容的总高度
        val scrollRange = editText.layout.height
        // 控件实际显示的高度
        val scrollExtent = (editText.height
                - editText.compoundPaddingTop
                - editText.compoundPaddingBottom)
        // 控件内容总高度与实际显示高度的差值
        val scrollDifference = scrollRange - scrollExtent
        return if (scrollDifference == 0) {
            false
        } else scrollY > 0 || scrollY < scrollDifference - 1
    }

    /**
     * 安装APK
     */
    fun installApk(context: Context, file: File?) {
        val intent = Intent(Intent.ACTION_VIEW)
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.N) {
            intent.setDataAndType(Uri.fromFile(file), "application/vnd.android.package-archive")
        } else {
            val apkUri = FileProvider.getUriForFile(context, context.packageName + ".fileprovider", file!!)
            intent.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
            intent.setDataAndType(apkUri, "application/vnd.android.package-archive")
        }
        context.startActivity(intent)
    }

    /**
     * 通知Media扫描
     *
     * @param context
     * @param file
     */
    fun notifyMediaScan(context: Context, file: File?) {
        val intent = Intent(Intent.ACTION_MEDIA_SCANNER_SCAN_FILE) //发送更新图片信息广播
        val uri = Uri.fromFile(file)
        intent.data = uri
        context.applicationContext.sendBroadcast(intent)
    }

    /**
     * 权限检查
     *
     * @param context     上下文
     * @param permissions 权限数组
     * @return 返回检查结果
     */
    @JvmOverloads
    fun checkPermissions(context: Context = ContextHelper.context, vararg permissions: String?): Boolean {
        if (permissions.isEmpty()) {
            return true
        }
        var isGranted = true
        permissions.forEach {
            isGranted = ActivityCompat.checkSelfPermission(context, it!!) == PackageManager.PERMISSION_GRANTED
            if (!isGranted) {
                return@forEach
            }
        }
        return isGranted
    }

    private var lastClickTime: Long = 0

    //这里设置的时间间隔是2s-判断是否频繁点击
    val isFastClick: Boolean
        get() = isFastClick(1000)

    /**
     * @param intervalTime 时间间隔
     * @return
     */
    fun isFastClick(intervalTime: Int): Boolean { //判断是否频繁点击
        val time = System.currentTimeMillis()
        val timeD = time - lastClickTime
        if (timeD in 1 until intervalTime) { // [,)
            return true
        }
        lastClickTime = time
        return false
    }

    /**
     * 是否主进程
     *
     * @param context
     * @return
     */
    @JvmOverloads
    fun isMainProcess(context: Context = ContextHelper.context): Boolean {
        val am = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val processInfos = am.runningAppProcesses
        val mainProcessName = context.packageName
        val myPid = Process.myPid()
        processInfos.forEach {
            if (it.pid == myPid && mainProcessName == it.processName) {
                return true
            }
        }
        return false
    }

    /**
     * 是否主线程
     *
     * @return
     */
    val isMainThread: Boolean
        get() = Looper.getMainLooper().thread === Thread.currentThread()

    /**
     * 获取系统基带
     *
     * @return
     */
    @get:SuppressLint("PrivateApi")
    val systemBaseband: String
        get() {
            try {
                val cl = Class.forName("android.os.SystemProperties")
                val invoker = cl.newInstance()
                val m = cl.getMethod("get", String::class.java, String::class.java)
                val result = m.invoke(invoker, "gsm.version.baseband", "")
                return result as String
            } catch (e: Exception) {
                e.printStackTrace()
            }
            return ""
        }

    /**
     * 判断是否打开了允许虚拟位置
     */
    fun isAllowMockLocation(context: Activity): Boolean {
        var isOpen = Settings.Secure.getInt(context.contentResolver,
                Settings.Secure.ALLOW_MOCK_LOCATION, 0) != 0
        /**
         * 该判断API是androidM以下的API,由于Android M中已经没有了关闭允许模拟位置的入口,
         * 所以这里一旦检测到开启了模拟位置,并且是android M以上,则默认设置为未有开启模拟位置
         */
        if (isOpen && Build.VERSION.SDK_INT > Build.VERSION_CODES.LOLLIPOP_MR1) {
            isOpen = false
        }
        return isOpen
    }

    /**
     * 获取进程号对应的进程名
     *
     * @param pid 进程号
     * @return 进程名
     */
    fun getProcessName(pid: Int): String? {
        var reader: BufferedReader? = null
        try {
            reader = BufferedReader(FileReader("/proc/$pid/cmdline"))
            var processName = reader.readLine()
            if (!TextUtils.isEmpty(processName)) {
                processName = processName.trim { it <= ' ' }
            }
            return processName
        } catch (throwable: Throwable) {
            throwable.printStackTrace()
        } finally {
            try {
                reader?.close()
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
        return null
    }

    /**
     * 获取服务是否开启
     */
    fun isServiceRunning(context: Context, className: String): Boolean {
        val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        // 获取正在运行的服务
        val runningServices = activityManager.getRunningServices(1000) //maxNum 返回正在运行的服务的上限个数,最多返回多少个服务
        for (runningServiceInfo in runningServices) {
            val service = runningServiceInfo.service
            // 获取正在运行的服务的全类名
            val className2 = service.className
            // 将获取到的正在运行的服务的全类名和传递过来的服务的全类名比较,
            // 一致表示服务正在运行返回true,不一致表示服务没有运行返回false
            return TextUtils.equals(className, className2)
        }
        return false
    }

    /**
     * 唤醒屏幕并解锁
     *
     * @param context
     */
    @SuppressLint("InvalidWakeLockTag", "WakelockTimeout")
    @RequiresPermission(Manifest.permission.DISABLE_KEYGUARD)
    fun wakeUpAndUnlock(context: Context) {
        try {
            val km = context.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
            val kl = km.newKeyguardLock("unLock")
            // 解锁
            kl.disableKeyguard()
            // 获取电源管理器对象
            val pm = context.getSystemService(Context.POWER_SERVICE) as PowerManager
            // 获取PowerManager.WakeLock对象,后面的参数|表示同时传入两个值,最后的是LogCat里用的Tag
            val wl = pm.newWakeLock(PowerManager.ACQUIRE_CAUSES_WAKEUP or PowerManager.SCREEN_DIM_WAKE_LOCK, "bright")
            // 点亮屏幕
            wl.acquire()
            // 释放
            wl.release()
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * 获取UDID
     */
    const val PREFS_FILE = "gank_device_id.xml"
    const val PREFS_DEVICE_ID = "gank_device_id"
    var uuid: String? = null

    @SuppressLint("InvalidWakeLockTag", "HardwareIds")
    @RequiresPermission(Manifest.permission.READ_PHONE_STATE)
    fun getUUID(context: Context): String? {
        if (uuid == null) {
            synchronized(PREFS_FILE) {
                if (uuid == null) {
                    val prefs = context.applicationContext.getSharedPreferences(PREFS_FILE, Context.MODE_PRIVATE)
                    val id = prefs.getString(PREFS_DEVICE_ID, null)
                    if (id != null) {
                        uuid = id
                    } else {
                        val androidId = Settings.Secure.getString(context.applicationContext.contentResolver, Settings.Secure.ANDROID_ID)
                        uuid = try {
                            if ("9774d56d682e549c" != androidId) {
                                UUID.nameUUIDFromBytes(androidId.toByteArray(charset("utf8"))).toString()
                            } else {
                                if (ActivityCompat.checkSelfPermission(context, Manifest.permission.READ_PHONE_STATE) == PackageManager.PERMISSION_GRANTED) {
                                    val deviceId = (context.applicationContext.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager).deviceId
                                    if (deviceId != null) UUID.nameUUIDFromBytes(deviceId.toByteArray(charset("utf8"))).toString() else UUID.randomUUID().toString()
                                } else {
                                    UUID.randomUUID().toString()
                                }
                            }
                        } catch (e: UnsupportedEncodingException) {
                            throw RuntimeException(e)
                        }
                        prefs.edit().putString(PREFS_DEVICE_ID, uuid).commit()
                    }
                }
            }
        }
        return uuid
    }

    fun getLauncherActivityByPkg(packagename: String?): String? {
        // 通过包名获取此APP详细信息，包括Activities、services、versioncode、name等等
        var packageinfo: PackageInfo? = null
        try {
            packageinfo = context.packageManager.getPackageInfo(packagename!!, 0)
        } catch (e: PackageManager.NameNotFoundException) {
            e.printStackTrace()
        }
        if (packageinfo == null) {
            return null
        }

        // 创建一个类别为CATEGORY_LAUNCHER的该包名的Intent
        val resolveIntent = Intent(Intent.ACTION_MAIN, null)
        resolveIntent.addCategory(Intent.CATEGORY_LAUNCHER)
        resolveIntent.setPackage(packageinfo.packageName)

        // 通过getPackageManager()的queryIntentActivities方法遍历
        val resolveinfoList = context.packageManager
                .queryIntentActivities(resolveIntent, 0)
        val resolveinfo = resolveinfoList.iterator().next()
        if (resolveinfo != null) {
            // packagename = 参数packname
//            val packageName = resolveinfo.activityInfo.packageName
            // 这个就是我们要找的该APP的LAUNCHER的Activity[组织形式：packagename.mainActivityname]
            return resolveinfo.activityInfo.name
        }
        return null
    }
}